includes/plugins/class-woocommerce-subscriptions.php

```diff
@@ -0,0 +1,113 @@
+ */
+
+namespace Newspack;
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Main class.
+ */
+class WooCommerce_Subscriptions {
+	/**
+	 * Initialize hooks and filters.
+	 */
+	public static function init() {
+		add_filter( 'woocommerce_subscription_settings', [ __CLASS__, 'add_on_hold_duration_setting' ], 11, 1 );
+		add_filter( 'wcs_default_retry_rules', [ __CLASS__, 'maybe_apply_on_hold_duration_rule' ], 99, 1 );
+	}
+
+	/**
+	 * Check if WooCommerce Subscriptions is active.
+	 *
+	 * @return bool
+	 */
+	public static function is_active() {
+		return class_exists( 'WC_Subscriptions' );
+	}
+
+	/**
+	 * Add on-hold duration setting.
+	 *
+	 * @param array $settings Subscription settings.
+	 *
+	 * @return array
+	 */
+	public static function add_on_hold_duration_setting( $settings ) {
+		if ( self::is_active() ) {
+			return array_merge(
+				$settings,
+				[
+					[
+						'name' => __( 'Newspack Subscriptions Settings', 'newspack-plugin' ),
+						'type' => 'title',
+						'desc' => __( 'Subscriptions settings added by Newspack.', 'newspack-plugin' ),
+						'id'   => 'newspack_subscriptions_options',
+					],
+					[
+						'name'              => __( 'On-hold Duration', 'newspack-plugin' ),
+						'id'                => 'newspack_subscriptions_on_hold_duration',
+						'css'               => 'max-width:80px;',
+						'value'             => self::get_on_hold_duration(),
+						'type'              => 'number',
+						'custom_attributes' => array(
+							'min'  => 0,
+							'step' => 1,
+						),
+						'desc'              => sprintf(
+							// Translators: %s is a line break.
+							__(
+								'Set the number of days a subscription remains in the On-Hold status after all automatic payment retries have failed.%sDuring this period, subscribers can update their payment details to restore the subscription at their current price. Once this time expires, the subscription will automatically transition to expired.',
+								'newspack-plugin'
+							),
+							'<br>'
+						),
+					],
+					[
+						'type' => 'sectionend',
+						'id'   => 'newspack_subscriptions_options',
+					],
+				]
+			);
+		}
+		return $settings;
+	}
+
+	/**
+	 * Get on-hold duration. Defaults to 0.
+	 *
+	 * @return int
+	 */
+	public static function get_on_hold_duration() {
+		return absint( get_option( 'newspack_subscriptions_on_hold_duration', 0 ) );
+	}
+
+	/**
+	 * Conditionally adds on-hold duration rule to retry rules.
+	 *
+	 * @param array $retry_rules Subscriptions retry rules.
+	 */
+	public static function maybe_apply_on_hold_duration_rule( $retry_rules ) {
+		if ( self::is_active() && count( $retry_rules ) > 0 ) {
+			$on_hold_duration = self::get_on_hold_duration();
+			if ( 0 < $on_hold_duration ) {
+				$retry_rules[] = [
+					'retry_after_interval'            => $on_hold_duration * DAY_IN_SECONDS,
+					'status_to_apply_to_order'        => 'pending',
+					'status_to_apply_to_subscription' => 'on-hold',
+				];
+			}
+			$retry_rules[] = [
```

Comment:
have you considered simply adding one rule with `$on_hold_duration` as interval and `expired` as status, instead of two rules?

And if `$on_hold_duration` is zero, change the status of the last rule?

I ask this because I think it would look a little cleaner in the admin. In my case, my on hold period was 3 days:

![image](https://github.com/user-attachments/assets/ad6b8a0d-2f7e-47df-8772-105412270ac5)

It's probably because I'm rushing the scheduled actions, but the last action shows out of the order, as if it happened before. Also, there is one additional attempt to process the payment right after one failed.

Reply:
> have you considered simply adding one rule with $on_hold_duration as interval and expired as status, instead of two rules?

The problem with this is these retry rules are parsed at the time of failed payment before the actual retry happens which means the status for a given retry rule will be applied at the time the rule is created and before the retry is actually triggered. So adding a single rule with `$on_hold_duration` as interval and `expired` as status would result in the subscription being set to `expired` on the fifth retry with the sixth retry attempt doomed to fail since the subscription has already been set to `expired`.

I hope this makes sense. Let me know if you'd like to huddle to talk it out.

Reply:
Ah, gotcha, that makes sense